solutions/devsprint-1/DeliveryAppChallenge/Service/DeliveryApi.swift

```diff
@@ -7,25 +7,74 @@
 
 import Foundation
 
-struct DeliveryApi {
+// MARK: - DeliveryApiError
 
-    func fetchRestaurants(_ completion: ([String]) -> Void) {
+enum DeliveryApiError: Error {
+    case invalidURL
+    case decodificationError
+    case responseError
+}
 
-        completion(["Restaurant 1", "Restaurant 2", "Restaurant 3"])
-    }
+// MARK: - DeliveryApiProtocol
 
-    func searchAddresses(_ completion: ([String]) -> Void) {
+protocol DeliveryApiProtocol {
+    func fetchRestaurants(_ completion: @escaping (Result<[Restaurant], DeliveryApiError>) -> Void)
+}
+
+// MARK: - DeliveryApi
 
+struct DeliveryApi {
+    func searchAddresses(_ completion: ([String]) -> Void) {
         completion(["Address 1", "Address 2", "Address 3"])
     }
 
     func fetchRestaurantDetails(_ completion: (String) -> Void) {
-
         completion("Restaurant Details")
     }
 
     func fetchMenuItem(_ completion: (String) -> Void) {
-
         completion("Menu Item")
     }
 }
+
+// MARK: - DeliveryApi
+
+extension DeliveryApi: DeliveryApiProtocol {
+    func fetchRestaurants(_ completion: @escaping (Result<[Restaurant], DeliveryApiError>) -> Void) {
+        guard let url = URL(string: "https://raw.githubusercontent.com/devpass-tech/challenge-delivery-app/main/api/home_restaurant_list.json") else {
+            return completion(.failure(.invalidURL))
+        }
+        let request = URLRequest(url: url)
+
+        let sessionConfig = URLSessionConfiguration.default
+        let session = URLSession(configuration: sessionConfig)
+
+        let task = session.dataTask(with: request) { data, response, error in
+            if let error = error {
+                print("==26===:  error", error)
+                return
+            }
+
+            guard
+                let response = response as? HTTPURLResponse,
+                response.statusCode == 200 else {
+                return completion(.failure(.responseError))
+            }
+
+            guard let data = data else {
+                return completion(.failure(.decodificationError))
+            }
+
+            guard let restaurants: [Restaurant] = data.jSONDecode(using: .convertFromSnakeCase) else {
+                return completion(.failure(.decodificationError))
+            }
+
+            restaurants.forEach { print($0) }
```

Comment:
Pode remover esse for também depois, ele está só dando print na primeira posição.